core/src/main/java/io/substrait/function/SimpleExtension.java

```diff
@@ -516,10 +524,43 @@ WindowFunctionVariant resolve(String uri, String name, String description) {
     }
   }
 
+  @JsonDeserialize(as = ImmutableSimpleExtension.Type.class)
+  @JsonSerialize(as = ImmutableSimpleExtension.Type.class)
+  @Value.Immutable
+  public abstract static class Type {
+    public abstract String name();
+
+    // TODO: Handle conversion of structure object to Named Struct representation
+    protected abstract Optional<Object> structure();
+
+    @Value.Default
+    public String uri() {
+      // we can't use null detection here since we initially construct this without a uri, then
+      // resolve later.
+      return "";
+    }
+
+    public Type resolve(String uri) {
+      return ImmutableSimpleExtension.Type.builder().name(name()).uri(uri).build();
+    }
+
+    public TypeAnchor getAnchor() {
+      return anchorSupplier.get();
+    }
+
+    private final Supplier<TypeAnchor> anchorSupplier =
+        Util.memoize(() -> TypeAnchor.of(uri(), name()));
+  }
+
   @JsonDeserialize(as = ImmutableSimpleExtension.FunctionSignatures.class)
   @JsonSerialize(as = ImmutableSimpleExtension.FunctionSignatures.class)
   @Value.Immutable
   public abstract static class FunctionSignatures {
+    // TODO: Rename to ExtensionSignatures ???
```

Comment:
I want to move away from FunctionSignature as this now has types in it as well, but I'm not sure if ExtensionSignature is a good name either. I'm open to suggestions if anyone has a better alternative.